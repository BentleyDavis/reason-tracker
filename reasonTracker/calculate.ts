import { Dictionary, getMany, getOne } from "@/utils/Dictionary";
import { Claim, isChildClaim } from "./types/Claim";
import { sortEdgeIdsLeavesToRoot } from "./sortLeavesToRoot";
import { claimChildrenIdsByParentId } from "./claimChildrenIdsByParentId";
import { Id, Item } from "./types/Item";
import { calculateConfidence } from "./calculateConfidence";
import { ChildScore } from "./types/Score";
import { calculateRelevance } from "./calculateRelevance";



export function calculate(claims: Dictionary<Claim>) {

    const childIndex = claimChildrenIdsByParentId(claims);

    // const confidences: Dictionary<ScoreWithDisplayData> = {};
    const scores: Dictionary<ChildScore<Id>> = {};

    // walk up the tree calculating the confidence of each claim
    for (const claimId of sortEdgeIdsLeavesToRoot(claims)) {
        const childrenIds = getOne(claimId, childIndex) || [];
        const claim = claims[claimId] as Claim;
        const childScores = getMany(childrenIds, scores);
        const score = calculateConfidence(childScores);
        const Relevance = calculateRelevance(childScores);
        if (isChildClaim(claim)) {
            scores[claimId] = {
                id: claimId,
                unitConfidence: score.unitConfidence,
                proMyParent: claim.pro,
                pctRelevantToMyParent: Relevance,
                affects: claim.affects
            }
        } else {
            // TODO: how to handle the root item?
        }
    }

    return scores

}


export type Confidence = Item<"Confidence", Id> & {
    confidence: number;
    treetype: "Node";
}


