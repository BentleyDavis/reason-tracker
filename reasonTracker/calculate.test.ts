import { calculate } from "./calculate";
import { simple } from "./examples/simple";
import { slightlyComplex } from "./examples/slightlyComplex";

describe("calculate", () => {

    it("should consistently solve Simple", () => {
        const claims = simple;

        const result = calculate(claims);

        expect(result).toMatchSnapshot();
    });

    it("should calculate scores for slightly complexity", () => {
        const claims = slightlyComplex;

        const result = calculate(claims);

        expect(result).toMatchSnapshot();
    });

});
