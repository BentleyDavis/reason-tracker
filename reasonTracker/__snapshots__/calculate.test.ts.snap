// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculate should calculate scores for slightly complexity 1`] = `
{
  "1": {
    "affects": "Confidence",
    "id": "1",
    "pctRelevantToMyParent": 1,
    "proMyParent": true,
    "unitConfidence": 1,
  },
  "2": {
    "affects": "Confidence",
    "id": "2",
    "pctRelevantToMyParent": 1,
    "proMyParent": false,
    "unitConfidence": 0.3333333333333333,
  },
  "2_1": {
    "affects": "Confidence",
    "id": "2_1",
    "pctRelevantToMyParent": 1,
    "proMyParent": true,
    "unitConfidence": 1,
  },
  "2_2": {
    "affects": "Confidence",
    "id": "2_2",
    "pctRelevantToMyParent": 1,
    "proMyParent": true,
    "unitConfidence": 1,
  },
  "2_3": {
    "affects": "Confidence",
    "id": "2_3",
    "pctRelevantToMyParent": 1,
    "proMyParent": false,
    "unitConfidence": 1,
  },
}
`;

exports[`calculate should consistently solve Simple 1`] = `
{
  "con1": {
    "affects": "Confidence",
    "id": "con1",
    "pctRelevantToMyParent": 1,
    "proMyParent": false,
    "unitConfidence": 1,
  },
  "pro1": {
    "affects": "Confidence",
    "id": "pro1",
    "pctRelevantToMyParent": 1,
    "proMyParent": true,
    "unitConfidence": 1,
  },
}
`;
